File: 234-PalindromeLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null)    return true;
        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        if (fast != null)
            slow = slow.next;
        ListNode p1 = head;
        ListNode p2 = reverseLinkedList(slow);
        while (p2 != null){
            if (p1.val != p2.val)
                return false;
            p1 = p1.next;
            p2 = p2.next;
        }
        return true;
    }
    
    public ListNode reverseLinkedList(ListNode head){
        if (head == null || head.next == null){
            return head;
        }
        ListNode pre = null;
        ListNode next = head;
        ListNode curr = head;
        while (curr != null){
            next = curr.next;
            curr.next = pre;
            pre = curr;
            curr = next;
        }
        return pre;
    }
}

File: cpp/234-PalindromeLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 //two situations:
 //1: 1->2->3->4->5
 //2: 1->2->3->4->5->6
 
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (!head || !(head->next)){
            return true;
        }
        ListNode* slow = head, *fast = head;
        while (fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        if (fast){
            slow = slow->next;
        }
        ListNode* rev = reverseList(slow);
        while (rev){
            if (head->val != rev->val){
                return false;
            }
            head = head->next;
            rev = rev->next;
        }
        return true;
    }

private:
    ListNode* reverseList(ListNode* head){
        ListNode* prev = NULL, *curr = head, *next = NULL;
        while (curr){
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
};